packages/adminrouter/extra/src/lib/auth/open.lua

```diff
@@ -22,33 +22,50 @@ else
 end
 
 
-local function validate_jwt_or_exit()
+local function validate_jwt_or_exit(proto)
     local uid, err = authcommon.validate_jwt(AUTH_TOKEN_VERIFICATION_KEY)
     if err ~= nil then
         if err == 401 then
-            return authcommon.exit_401("oauthjwt")
+            if proto == "grpc" then
+                return authcommon.exit_grpc(16, "access denied - not authenticated")
+            else
+                return authcommon.exit_401("oauthjwt")
+            end
         end
 
         -- Catch-all, normally not reached:
         ngx.log(ngx.ERR, "Unexpected result from validate_jwt()")
-        ngx.status = ngx.HTTP_INTERNAL_SERVER_ERROR
-        return ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
+        if proto == "grpc" then
+            return authcommon.exit_grpc(13, "internal error occurred")
+        else
+            ngx.status = ngx.HTTP_INTERNAL_SERVER_ERROR
+            return ngx.exit(ngx.HTTP_INTERNAL_SERVER_ERROR)
+        end
     end
     return uid
 end
 
 
-local function do_authn_and_authz_or_exit()
+local function do_authn_and_authz_or_exit_http()
+    -- Here, only do authentication, i.e. require
+    -- valid authentication token to be presented.
+    -- Downstream, this function can be replaced
+    -- with more complex business logic.
+    validate_jwt_or_exit("http")
+end
+
+
+local function do_authn_and_authz_or_exit_grpc()
     -- Here, only do authentication, i.e. require
     -- valid authentication token to be presented.
     -- Downstream, this function can be replaced
     -- with more complex business logic.
-    validate_jwt_or_exit()
+    validate_jwt_or_exit("grpc")
 end
 
 
 local function do_authn_or_exit(object)
```

Comment:
Maybe we should also need `do_authn_or_exit` for `grpc`? At least it will be used in EE.

Reply:
We do not use `do_authn_or_exit_grpc` at all, EE uses `res.do_authn_and_authz_or_exit_grpc`. Do you think there is a point in creating a function that is not going to be called?